nixos/modules/services/hardware/logiops.nix

```diff
@@ -0,0 +1,139 @@
+                invert: false;
+                target: false;
+            };
+            dpi: 1000;
+
+            buttons: (
+                {
+                    cid: 0xc3;
+                    action =
+                    {
+                        type: "Gestures";
+                        gestures: (
+                            {
+                                direction: "Up";
+                                mode: "OnRelease";
+                                action =
+                                {
+                                    type: "Keypress";
+                                    keys: ["KEY_UP"];
+                                };
+                            },
+                            {
+                                direction: "Down";
+                                mode: "OnRelease";
+                                action =
+                                {
+                                    type: "Keypress";
+                                    keys: ["KEY_DOWN"];
+                                };
+                            },
+                            {
+                                direction: "Left";
+                                mode: "OnRelease";
+                                action =
+                                {
+                                    type: "CycleDPI";
+                                    dpis: [400, 600, 800, 1000, 1200, 1400, 1600];
+                                };
+                            },
+                            {
+                                direction: "Right";
+                                mode: "OnRelease";
+                                action =
+                                {
+                                    type = "ToggleSmartshift";
+                                }
+                            },
+                            {
+                                direction: "None"
+                                mode: "NoPress"
+                            }
+                        );
+                    };
+                },
+                {
+                    cid: 0xc4;
+                    action =
+                    {
+                        type: "Keypress";
+                        keys: ["KEY_A"];
+                    };
+                }
+            );
+        }
+        );
+      '';
+    };
+
+    # settings = mkOption {
+    #   type = types.submodule {
+    #     # Declare that the settings option supports arbitrary format values, libconfig here
+    #     freeformType = settingsFormat.type;
+    #   };
+    #   default = {};
+    #   # Add upstream documentation to the settings description
+    #   description = ''
+    #     Configuration for `logiops`, see
+    #     <link xlink:href="https://github.com/PixlOne/logiops/wiki/Configuration"/>
+    #     for supported values.
+    #   '';
+    # };
+  };
+
+  config = mkIf cfg.enable {
+    systemd.services.logiops = {
```

Comment:
I think you should reuse the upstream `systemd` unit and then customize it. If you're not familiar how to do that please feel free to ask questions and I can go over in detail. From a high level you just add the package to `systemd.packages` and that will include any units in the resulting system. From there you can add customizations as needed. Be sure to keep the `wantedBy` directive because NixOS can't parse that out of upstream units.

You'll also want to patch the location of `/usr/bin/kill` in the upstream unit.

Reply:
I must admit I haven't the foggiest clue how I would reuse the upstream systemd unit and then customize it, but I agree that it is a better approach. I am very keen to learn, though!

[Initially, my derivation wouldn't build because the upstream unit file installation failed](https://github.com/NixOS/nixpkgs/issues/122208#issuecomment-840444771), so I looked around `nixpkgs` for what other packages had done and found this: https://github.com/NixOS/nixpkgs/blob/85784a66c8848de93c6ee3d0fb47ec425491f404/pkgs/servers/monitoring/icinga2/no-systemd-service.patch, and I copied that approach. Specifically, when I don't apply my CMake patch, the error I get is:
```
…
[ 93%] Building CXX object src/logid/CMakeFiles/logid.dir/util/worker_thread.cpp.o
[ 95%] Building CXX object src/logid/CMakeFiles/logid.dir/util/task.cpp.o
[ 96%] Building CXX object src/logid/CMakeFiles/logid.dir/util/thread.cpp.o
[ 98%] Building CXX object src/logid/CMakeFiles/logid.dir/util/ExceptionHandler.cpp.o
[100%] Linking CXX executable ../../logid
[100%] Built target logid
installing
install flags: SHELL=/nix/store/xvvgw9sb8wk6d2c0j3ybn7sll67s3s4z-bash-4.4-p23/bin/bash install
[100%] Built target logid
Install the project...
-- Install configuration: "Release"
-- Installing: /nix/store/vi59a2gy74cg81vd2smk920jiqz36nvl-logiops-0.2.3/bin/logid
CMake Error at src/logid/cmake_install.cmake:70 (file):
  file cannot create directory:
  /nix/store/71lqc2a8cslg4wxj6ypla7gvflphjhn0-systemd-247.6/lib/systemd/system.
  Maybe need administrative privileges.
Call Stack (most recent call first):
  cmake_install.cmake:47 (include)


make: *** [Makefile:105: install] Error 1
builder for '/nix/store/hjxy35w6xkwb5hivml15y9m7kwng0x90-logiops-0.2.3.drv' failed with exit code 2
cannot build derivation '/nix/store/diz9yg8vxxaisfdisiqxhpm5fazsx8pf-unit-logiops.service.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/x88mi1xav2csv6n5nzdf4xwc9hl8gj2f-system-units.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/i6s5n1700gcq8xlqxwj57cfabh3h74rn-etc.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/67l36cbh1gfxnh2xig67y3v5hmwm37dv-nixos-system-monarch-21.05.2549.a1007637cea.drv': 1 dependencies couldn't be built
error: build of '/nix/store/67l36cbh1gfxnh2xig67y3v5hmwm37dv-nixos-system-monarch-21.05.2549.a1007637cea.drv' failed
```

Reply:
@zeorin this looks like a "bug" (intended feature which was never properly implemented?) from upstream to me:

https://github.com/PixlOne/logiops/blob/6bb470000952405b6817032c43c7215f7d070428/src/logid/CMakeLists.txt#L92-L105

If you set the value of `SYSTEMD_SERVICES_INSTALL_DIR` then upstream doesn't do install the `.service` file, which seems wrong.

I would propose that you create a PR upstream to fix this with the following commit: https://github.com/aanderse/logiops/commit/c723d0fb6bcc647ee35fe3eeb728e21ba93e1c90

I tested the above commit with this commit on top of your PR in nixpkgs and it all built as expected with the desired outcome: https://github.com/zeorin/nixpkgs/compare/feature/logiops-init...aanderse:logiops?expand=1

Let me know if you have any questions or need a hand on this.